libyul/AST.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+This file is part of solidity.
+
+solidity is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+solidity is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/AST.h>
+#include <libyul/Exceptions.h>
+
+namespace solidity::yul
+{
+
+LiteralValue::LiteralValue(std::string _builtinStringLiteralValue):
+	m_numericValue(0),
```

Comment:
You don't seem to use the fact that it's an `optional`:
```suggestion
LiteralValue::LiteralValue(std::string _builtinStringLiteralValue):
	m_numericValue(std::nullopt),
```

With that you also don't really need the `m_builtinStringLiteralValue` flag.

BTW, class docstring says ``builtins - exceed the u256 word (32 bytes), in which case the value is set to `unlimited` ``. Did you meant that `m_builtinStringLiteralValue` is set to `true`?

Reply:
Also, in `SyntacticallyEqual::expressionEqual()` you compare `value()` of normal and unlimited strings this way:
```c++
return valid && _lhs.value.value() == _rhs.value.value();
```
If you set the value to zero for unlimited they will compare as equal to a string `""`, which does not seem like it would represent actual syntactical equality.